design/hotshard.md

```diff
@@ -0,0 +1,385 @@
+### Storage Server Impact
+
+- The storage servers hosting the hot shard become saturated with write traffic, causing their write queues to grow [13][14].
+
+- Storage server metrics report high `bytesInput` and increasing queue depths, which are monitored by Ratekeeper [13].
+
+- If the storage server's write queue size exceeds thresholds relative to available space and target ratios, it becomes the bottleneck for the entire cluster [14].
+
+### Cluster-Wide Rate Limiting
+
+- Ratekeeper continuously monitors storage server queues and bandwidth metrics to determine cluster-wide transaction rate limits [13][14].
+
+- When a storage server's write queue becomes excessive, Ratekeeper sets the limit reason to `storage_server_write_queue_size` and reduces the cluster-wide transaction rate proportionally [14][15].
+
+- This causes **all transactions across the cluster** to be throttled, even those not touching the hot shard, because the cluster cannot commit faster than its slowest storage server can durably persist data.
+
+- The rate limit is calculated as: `limitTps = min(actualTps * maxBytesPerSecond / inputRate, maxBytesPerSecond * MAX_TRANSACTIONS_PER_BYTE)`, ensuring the cluster doesn't overwhelm the saturated storage server [13].
+
+## Mitigation
+
+### Avoiding Hot Shards
+
+- Randomize key prefixes so consecutive writes land on different shard ranges. For example, hash user IDs or add a short random salt before the natural key. This way inserts will scatter instead of piling onto one shard.
+- If you need to store a counter, consider sharding them across N disjoint keys (e.g., counter/<bucket>/…) and aggregate in clients or background jobs. This keeps the per-key mutation rate below the commit proxy’s hot-shard throttle.
+- If you are storing append-only logs,  split them into multiple partitions (such as log/<partition>/<ts>), rotating partitions over time rather than funneling through a single key path.
+- Avoid “read-modify-write” cycles. Use FDB's atomic operations (like ATOMIC_ADD) when possible, and throttle/queue work in clients so they don’t stampede on that hot key.
+- For multi-tenant schemas, assign each tenant a disjoint key subspace.
+
+### Avoiding Read Hotspots
+
+  - Cache hot objects (in the application or via a memory service) instead of repeatedly issuing get for the same key every transaction; read-hot shards
+  often stem from the same value being fetched continuously.
+  - For range scans, fetch only the window you truly need—store derived indexes or “chunked” materializations so transactions avoid reading a large
+  contiguous range every time.
+  - If one small document is hammered by many readers, replicate it into multiple keys (e.g., per region or per hash bucket) and route readers to
+  different replicas based on a deterministic hash. Background jobs can keep replicas in sync.
+  - Stagger periodic polling across clients—add random jitter so metrics or heartbeat jobs don’t all read the same shard simultaneously.
```

Comment:
Maybe worth to add here, that depending on the use case, instead of polling you could use the watch feature of FDB.